### PR TITLE
fix(klaviyo): Ensure marketing opt-in triggers subscription

### PR DESCRIPTION
**Pull request by jawaria-sparksol**
### Summary
This pull request fixes a bug where users were not being subscribed to the Klaviyo marketing list after opting in.
### The Problem
The subscription job was not being triggered when a user opted into marketing emails via checkout, the newsletter form, or their account page.
### The Solution
The logic in the **marketing_opt_in_changed?** method has been corrected. It now ensures the subscription job is dispatched only when a user's accepted_marketing status is newly changed to true, fixing the issue where the process was previously blocked.
### How to Test
**Action:** Opt-in to marketing via the newsletter form, checkout, or the "My Account" page.
**Verification:** Confirm the user is added to the Klaviyo list VnA882. If double opt-in is enabled, check for the confirmation email first.

## Summary by CodeRabbit

* **New Features**
  * Automatically subscribes users to Klaviyo when they newly opt into marketing during profile updates; only triggers on real opt-in changes and avoids duplicate subscriptions.

* **Tests**
  * Added specs validating subscription job enqueueing for active integrations and non-enqueueing when missing/inactive.
  * Updated several service specs and test setups to explicitly opt users into email marketing for compatibility.

### CODE REVIEW

File: app/models/concerns/spree_klaviyo/user_methods.rb [hunk 1]
```diff
@@ -20,5 +22,20 @@ def fetch_klaviyo_profile(klaviyo_integration:)
...
+    private
+
+    def marketing_opt_in_changed?
+      return false if klaviyo_subscribed?
+
+      saved_change_to_accepted_marketing? && accepted_marketing?
+    end
+
+    def subscribe_to_klaviyo
+      klaviyo_integration = ::Spree::Integrations::Klaviyo.last
```
Comment (brozek95) [inline app/models/concerns/spree_klaviyo/user_methods.rb:35]: Don't we need to check if this integration is active?

Comment (jawaria-sparksol) [reply app/models/concerns/spree_klaviyo/user_methods.rb]: I have updated this to select active integration only.

File: spec/models/spree/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        it 'does not enqueue a SubscribeJob' do
+          expect {
+            user.send(:subscribe_to_klaviyo)
+          }.not_to have_enqueued_job(SpreeKlaviyo::SubscribeJob)
+        end
+      end
+    end
+
+    describe '#marketing_opt_in_changed? (private)' do
```
[the remark this answers is not in the document]
Comment (jawaria-sparksol) [reply spec/models/spree/user_spec.rb:49]: Thanks for the suggestion! I initially tried the approach of testing the after_commit callback with real user updates (which would indeed be better), but unfortunately the **accepted_marketing** attribute doesn't exist on the Spree user model in this setup.
When I attempted to use **user.update!(accepted_marketing: true)**, it threw:
`The marketing_opt_in_changed? method calls saved_change_to_accepted_marketing? and accepted_marketing? which are also undefined methods on the user model.`
So while testing the actual callback behavior would be ideal, the stubbing approach is the most appropriate here since:
- It tests the business logic of marketing_opt_in_changed? in isolation
- It covers all edge cases without requiring non-existent infrastructure
- It's the only viable approach given the current user model structure
- If we want to implement the end-to-end testing approach you suggested, we'd first need to add the accepted_marketing attribute to the user model (likely via a migration or store_accessor).

File: app/models/concerns/spree_klaviyo/user_methods.rb [hunk 3]
```diff
@@ -20,5 +22,20 @@ def fetch_klaviyo_profile(klaviyo_integration:)
 
       SpreeKlaviyo::FetchProfileJob.perform_later(klaviyo_integration.id, id)
     end
+
+    private
+
+    def marketing_opt_in_changed?
+      return false if klaviyo_subscribed?
+
+      saved_change_to_accepted_marketing? && accepted_marketing?
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply app/models/concerns/spree_klaviyo/user_methods.rb:31]: https://github.com/spree/spree/blob/main/core/db/migrate/20250207084000_add_accepts_email_marketing_field_to_spree_users_table.rb#L3

File: spec/models/spree/user_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jawaria-sparksol) [reply spec/models/spree/user_spec.rb:49]: Updated!

File: spec/models/spree/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      before do
+        # user.klaviyo_subscribed = false
```
[the remark this answers is not in the document]
Comment (jawaria-sparksol) [reply spec/models/spree/user_spec.rb:51]: Removed.

File: spec/models/spree/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        end
+
+        it 'returns false when already Klaviyo-subscribed even if flag changed to true' do
+          u = create(:user, accepts_email_marketing: false)
+          u.klaviyo_subscribed = true
+          u.update!(accepts_email_marketing: true)
+          expect(u.accepts_email_marketing).to be(true)
+          expect(u.send(:marketing_opt_in_changed?)).to be(false)
+        end
+      end
```
[the remark this answers is not in the document]
Comment (jawaria-sparksol) [reply spec/models/spree/user_spec.rb:84]: @brozek95 should I revert my specs to previous state as that is also suggested in this comment.

[the remark this answers is not in the document]
Comment (brozek95) [reply spec/models/spree/user_spec.rb:84]: I just want you to check if ai is right here.

[the remark this answers is not in the document]
Comment (jawaria-sparksol) [reply spec/models/spree/user_spec.rb:84]: Thanks for the suggestions. I’d like to keep the current tests as-is for these reasons:

- The canonical field in Spree core is accepts_email_marketing, not accepted_marketing. Our tests validate that path end-to-end (dirty tracking on the same instance), which stub-only tests can’t guarantee. See Spree’s migration adding this field: https://github.com/spree/spree/blob/main/core/db/migrate/20250207084000_add_accepts_email_marketing_field_to_spree_users_table.rb#L3
- We already guard the integration-style tests with a skip when the column isn’t present, so they won’t break older Spree versions.
- Setting klaviyo_subscribed directly is correct in this gem: it’s a store_accessor, and we’re verifying our own concern’s behavior, not persistence.

Given the above, I don’t see a need to revert to stub-only tests.

[the remark this answers is not in the document]
Comment (brozek95) [reply spec/models/spree/user_spec.rb]: Sorry, I didn't noticed the AI wants to revert to "mock-all" pattern.

**Review by brozek95 — APPROVED** [review]

File: spec/services/spree_klaviyo/create_or_update_profile_spec.rb [hunk 6]
```diff
@@ -3,6 +3,10 @@
 RSpec.describe SpreeKlaviyo::CreateOrUpdateProfile do
   subject { described_class.call(klaviyo_integration: klaviyo_integration, user: user) }
 
+  before do
```
Comment (damianlegawiec) [inline spec/services/spree_klaviyo/create_or_update_profile_spec.rb:6]: @jawaria-sparksol please create user with `accepts_email_marketing` true instead of this

File: spec/services/spree_klaviyo/fetch_profile_spec.rb [hunk 7]
```diff
@@ -5,6 +5,10 @@
 
   let(:user) { create(:user) }
 
+  before do
```
Comment (damianlegawiec) [inline spec/services/spree_klaviyo/fetch_profile_spec.rb:8]: @jawaria-sparksol same here

**Review by damianlegawiec — APPROVED** [review]
